solutions/platinum/leetcode-149.mdx

```diff
@@ -84,4 +85,104 @@ class Solution {
 ```
 
 </JavaSection>
+<PySection>
+
+```py
+class Solution:
+	def maxPoints(self, points: List[List[int]]) -> int:
+		n = len(points)
+		if n <= 2:
+			return n
+
+		ans = 0
+		for i in range(n):
+			for j in range(i + 1, n):
+				p = 2  # the 2 points are collinear with themselves
+				for k in range(j + 1, n):
+					dx1 = points[i][0] - points[k][0]
+					dx2 = points[j][0] - points[i][0]
+					dy1 = points[i][1] - points[k][1]
+					dy2 = points[j][1] - points[i][1]
+
+					# Check if dy1 / dx1 = dy2 / dx2
+					# Which is the same as: dy1 * dx2 = dy2 * dx1
+					if dy1 * dx2 == dy2 * dx1:
+						p += 1
+
+				ans = max(ans, p)
+
+		return ans
+```
+
+</PySection>
+</LanguageSection>
+
+## Efficient Solution
+
+We can take a point $A$ and compute the slopes of all lines connecting $A$ to the other points. Two different points lie on the same line if they share the same slope. In this way, we can determine the maximum number of points lying on the same straight line wrt $A$.
+
+We perform this for all other points.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(n^2)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+class Solution {
+  public:
+	int maxPoints(vector<vector<int>> &points) {
+		int n = (int)points.size();
+		int res = 0;
+		for (int i = 0; i < n; i++) {
+			unordered_map<float, int> umap;
```

Comment:
1. is unordered map necessary
2. why not name it slope_count like you did w/ python

Reply:
1. Yes, because of its constant insertion operation.

Reply:
2. Done the changes

Reply:
wait it like actually TLEs w/ normal map?

Reply:
No, since with a normal map the insertion operation is costly that's why it is better to use an unordered map for efficiency.

Reply:
ok just making sure you actually tried right
because umap constant factor is stupid bad

Reply:
Yes the constant factor is worse. 
At first I did thought of using a normal map but since we don't have to worry about sorted elements that's why I excluded that idea.